### PR TITLE
fix(eslint-plugin): [no-unsafe-assignment] be more specific about error types

### LINKED ISSUE #8231: Bug: [no-unsafe-assignment] Differentiate a types-error any from a "true" any
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.ts&code=DYUwLgBAbghsCuIBcEB28C2AjEAnA3AFCGwIgQC8EAzgA4D29A1gJ75A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolp9kySgHMmAW0Qt0URNGgdokcGAC%2BITUA&tsconfig=&tokens=false

### Repro Code

```TypeScript
let value: number;

value = spooky;
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unsafe-assignment": "error"
  }
}
```

### tsconfig

_No response_

### Expected Result

Users should be told that the type is an "error" variant of `any`. It's confusing to have the same report show up for both a "real" `any` and an error in type checking.

### Actual Result

The same error as if we'd added a `declare const spooky: any;`:

```plaintext
Unsafe assignment of an `any` value.
```

### Additional Info

Although the type's `typeToString()` returns `"any"`, its intrinsic name is `"error"`. See the _Types_ panel of the linked playground.

Error types in type checking are often a symptom of misconfiguration around TypeScript and/or typescript-eslint. [Example StackOverflow question around an `any` in a lint rule](https://stackoverflow.com/questions/77659106/overcome-eslints-redundant-type-constituents-for-disposable-interface).

### PR DESCRIPTION
**Pull request by tobySolutions**
## PR Checklist

- [x] Addresses an existing open issue: fixes #8231 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

### CODE REVIEW

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
A great start, thanks again for working on this with me!

Requesting changes mostly around adding in more testing. A couple small refactors too. But the functionality looks great! ✨

**Comment by JoshuaKGoldberg** [comment]
👋 @tobySolutions ping, just checking in - is this something you still have the time and energy for? It's still in draft & there are some CI ❌s so we haven't looked at it.

**Comment by tobySolutions** [comment]
> 👋 @tobySolutions ping, just checking in - is this something you still have the time and energy for? It's still in draft & there are some CI ❌s so we haven't looked at it.

Definitely Josh! I will resume work on this in a bit. Thank you so very much for checking in! 🙌

File: yarn.lock [hunk 1]
```diff
@@ -19699,11 +19699,11 @@ __metadata:
 
 "typescript@patch:typescript@npm%3A5.4.5#~builtin<compat/typescript>":
   version: 5.4.5
-  resolution: "typescript@patch:typescript@npm%3A5.4.5#~builtin<compat/typescript>::version=5.4.5&hash=5adc0c"
+  resolution: "typescript@patch:typescript@npm%3A5.4.5#~builtin<compat/typescript>::version=5.4.5&hash=e012d7"
   bin:
     tsc: bin/tsc
     tsserver: bin/tsserver
-  checksum: d59e26e74f6b444517d0ba16e0ee16e75c652c2b49a59f2ebdbeb16647a855fd50c7fc786a58987e45f03bce0677092e2dd3333649fd53b11d0b0d271455837c
+  checksum: 2373c693f3b328f3b2387c3efafe6d257b057a142f9a79291854b14ff4d5367d3d730810aee981726b677ae0fd8329b23309da3b6aaab8263dbdccf1da07a3ba
```
Comment (JoshuaKGoldberg) [inline yarn.lock:19706]: Unrelated changes? You'll want to revert these.

File: packages/type-utils/src/predicates.ts [hunk 2]
```diff
@@ -95,6 +95,16 @@ export function isTypeAnyType(type: ts.Type): boolean {
   return false;
 }
 
+/**
+ * @returns true if there is the type is an `error` type
+ */
+export function isTypeErrorType(type: ts.Type): boolean {
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/type-utils/src/predicates.ts:101]: Bumping this thread: you can delete this function. `tsutils.isIntrinsicErrorType` is already available to do this same thing.

File: yarn.lock (hunk 1 above)
Comment (tobySolutions) [reply yarn.lock:19706]: As in revert all my changes?

File: packages/eslint-plugin/src/rules/no-unsafe-assignment.ts [hunk 3]
```diff
@@ -37,17 +38,18 @@ export default createRule({
       requiresTypeChecking: true,
     },
     messages: {
-      anyAssignment: 'Unsafe assignment of an `any` value.',
+      anyAssignment: 'Unsafe assignment of an {{sender}} typed value.',
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-unsafe-assignment.ts:41]: > `{{ sender }} typed`value.

There's no need to add the _"typed"_ for _all_ error reports. Just the new one. Existing error reports don't need to say `... an any typed value`.

```suggestion
      anyAssignment: 'Unsafe assignment of an {{sender}} value.',
```

Suggestion: move the "typed" into the `data` passed by the `createDataFromSenderType` function?

File: packages/eslint-plugin/src/rules/no-unsafe-assignment.ts [hunk 4]
```diff
@@ -37,17 +38,18 @@ export default createRule({
...
       anyAssignmentThis: [
-        'Unsafe assignment of an `any` value. `this` is typed as `any`.',
+        'Unsafe assignment of an {{sender}}  value. `this` is typed as `any`.',
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-unsafe-assignment.ts:43]: [Bug] Looks like a bad find-and-replace, two spaces? Here and elsewhere:

```suggestion
        'Unsafe assignment of an {{sender}} value. `this` is typed as `any`.',
```

File: yarn.lock (hunk 1 above)
Comment (tobySolutions) [reply yarn.lock:19706]: Done.

File: packages/eslint-plugin/tests/rules/no-unsafe-assignment.test.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/no-unsafe-assignment.test.ts:1]: [Testing] Now that the errors include `data`, you'll want to include that in the tests.

```diff
- errors: [{ messageId: 'anyAssignment' }],
+ errors: [{ data: 'any', messageId: 'anyAssignment' }],
```

Without including `data` in the errors, the tests aren't really testing everything we need them to. This PR adds in new logic around the `data` that should be tested!

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Progress! 🚀

File: packages/eslint-plugin/src/rules/no-unsafe-assignment.ts (hunk 3 above)
Comment (tobySolutions) [reply packages/eslint-plugin/src/rules/no-unsafe-assignment.ts:41]: Done.

File: packages/eslint-plugin/src/rules/no-unsafe-assignment.ts (hunk 4 above)
Comment (tobySolutions) [reply packages/eslint-plugin/src/rules/no-unsafe-assignment.ts:43]: Done.

File: packages/eslint-plugin/tests/rules/no-unsafe-assignment.test.ts
Comment (tobySolutions) [reply packages/eslint-plugin/tests/rules/no-unsafe-assignment.test.ts:1]: Do you mean something like this @JoshuaKGoldberg?

```
  errors: [
        { messageId: 'unsafeArrayPatternFromTuple', data: { sender: 'any', receiver: 'any'}},
        { messageId: 'anyAssignment', data: { sender: 'any', receiver: 'any'}},
      ],
```

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/tests/rules/no-unsafe-assignment.test.ts:1]: Yup, exactly!

Comment (tobySolutions) [reply packages/eslint-plugin/tests/rules/no-unsafe-assignment.test.ts:1]: Done

File: packages/eslint-plugin/src/rules/no-unsafe-assignment.ts [hunk 5]
```diff
@@ -298,8 +304,8 @@ export default createRule({
         node: reportingNode,
         messageId: 'unsafeAssignment',
         data: {
-          sender: checker.typeToString(sender),
-          receiver: checker.typeToString(receiver),
+          sender: '`' + checker.typeToString(sender) + '`',
+          receiver: '`' + checker.typeToString(receiver) + '`',
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-unsafe-assignment.ts:308]: [Question] Is there a reason this `data` object isn't using `creatDataFromSenderType`? I don't remember 🤔

File: packages/eslint-plugin/tests/rules/no-unsafe-assignment.test.ts [hunk 6]
```diff
@@ -195,7 +196,50 @@ class Foo {
       `,
       errors: [{ messageId: 'anyAssignment' }],
     },
+    {
+      code: `
+const [x] = spooky;
+      `,
+      errors: [{ messageId: 'anyAssignment' }],
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/no-unsafe-assignment.test.ts:203]: [Testing] These first two added test cases should have the `data` too. Every test added in this PR is something where the data having `error` or `any` is important enough to test, I'd think.

File: packages/eslint-plugin/src/rules/no-unsafe-assignment.ts (hunk 5 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/no-unsafe-assignment.ts:308]: @tobySolutions you marked this as resolved but I don't know what the answer is 😂 what is the resolution here? What's up?

**Review by JoshuaKGoldberg — APPROVED** [review]
🙌 Great! Thanks for working with me on this @tobySolutions. I'm excited to see a better error message for `no-unsafe-assignment` on production soon!

Only one change request left from me, to add a bit more to the tests. It's a pretty small thing - so marking as `1 approval` to let someone else from the team chime in if the want.


Cartoon cat happily jumping for joy. Caption: "YAY!!!"

**Comment by tobySolutions** [comment]
> 🙌 Great! Thanks for working with me on this @tobySolutions. I'm excited to see a better error message for `no-unsafe-assignment` on production soon!
> 
> Only one change request left from me, to add a bit more to the tests. It's a pretty small thing - so marking as `1 approval` to let someone else from the team chime in if the want.
> 
> Cartoon cat happily jumping for joy. Caption: "YAY!!!" [ Cartoon cat happily jumping for joy. Caption: "YAY!!!" ](https://private-user-images.githubusercontent.com/3335181/335918460-807e4608-7f9f-4a05-9941-0d8951685460.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTczODA4OTAsIm5iZiI6MTcxNzM4MDU5MCwicGF0aCI6Ii8zMzM1MTgxLzMzNTkxODQ2MC04MDdlNDYwOC03ZjlmLTRhMDUtOTk0MS0wZDg5NTE2ODU0NjAuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDNUMDIwOTUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2Q2NWI3YjkxZjY0NWY5M2QzYzMzMDVjYzU2NzI2ZWYyM2NmODI4ZjIzZDJlMjJjOGQ3YzcyYmFiYzBkZjAzNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.QoL67Sul87QWOtIehXb_YdhwzbnGKCAdtJc1CNu5EvI) [ ](https://private-user-images.githubusercontent.com/3335181/335918460-807e4608-7f9f-4a05-9941-0d8951685460.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTczODA4OTAsIm5iZiI6MTcxNzM4MDU5MCwicGF0aCI6Ii8zMzM1MTgxLzMzNTkxODQ2MC04MDdlNDYwOC03ZjlmLTRhMDUtOTk0MS0wZDg5NTE2ODU0NjAuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDNUMDIwOTUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2Q2NWI3YjkxZjY0NWY5M2QzYzMzMDVjYzU2NzI2ZWYyM2NmODI4ZjIzZDJlMjJjOGQ3YzcyYmFiYzBkZjAzNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.QoL67Sul87QWOtIehXb_YdhwzbnGKCAdtJc1CNu5EvI)

Yay!!!

**Comment by Josh-Cena** [comment]
If it only changes one rule, then it doesn't fix #8231, only partially

**Comment by JoshuaKGoldberg** [comment]
Yup - I'm about to post an edit splitting the issue up 😄

**Comment by Josh-Cena** [comment]
What about other `no-unsafe` rules?

**Comment by JoshuaKGoldberg** [comment]
🙌 Filed!

**Review by JoshuaKGoldberg — APPROVED** [review]
This is wonderful, thanks @tobySolutions!

We've got a bunch of issues filed to apply similar improvements to other `no-unsafe-*` rules. I think once we've gone through those we'll be well-positioned to review our messaging around "error" types and maybe iterate further on how they're explained to users. But that's a battle for another day.
 
Nicholas Cage happily smiling with long hair flowing in a breeze. His head fades into a cosmic galaxy kind of fly-through

**Comment by tobySolutions** [comment]
> This is wonderful, thanks @tobySolutions!
> 
> We've got a bunch of issues filed to apply similar improvements to other `no-unsafe-*` rules. I think once we've gone through those we'll be well-positioned to review our messaging around "error" types and maybe iterate further on how they're explained to users. But that's a battle for another day.
> 
> Nicholas Cage happily smiling with long hair flowing in a breeze. His head fades into a cosmic galaxy kind of fly-through [ Nicholas Cage happily smiling with long hair flowing in a breeze. His head fades into a cosmic galaxy kind of fly-through ](https://private-user-images.githubusercontent.com/3335181/337011941-cdbf7f09-e277-4a07-9519-db2454e2a3a6.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc2MjYzNjQsIm5iZiI6MTcxNzYyNjA2NCwicGF0aCI6Ii8zMzM1MTgxLzMzNzAxMTk0MS1jZGJmN2YwOS1lMjc3LTRhMDctOTUxOS1kYjI0NTRlMmEzYTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDVUMjIyMTA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjVlYzliMTkxYzM4YzhlMDM5MDM0MmZiNWUzMjEzMTY4YzRkZDc1ZTViMTQ3NjUxNzJiMzk3NTFjOWE5ZTU4MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.4eKs9gOHWO6Pv2GfxXe6z3LrWMiEUF9wbvO_p7fo9DE) [ ](https://private-user-images.githubusercontent.com/3335181/337011941-cdbf7f09-e277-4a07-9519-db2454e2a3a6.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc2MjYzNjQsIm5iZiI6MTcxNzYyNjA2NCwicGF0aCI6Ii8zMzM1MTgxLzMzNzAxMTk0MS1jZGJmN2YwOS1lMjc3LTRhMDctOTUxOS1kYjI0NTRlMmEzYTYuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDVUMjIyMTA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjVlYzliMTkxYzM4YzhlMDM5MDM0MmZiNWUzMjEzMTY4YzRkZDc1ZTViMTQ3NjUxNzJiMzk3NTFjOWE5ZTU4MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.4eKs9gOHWO6Pv2GfxXe6z3LrWMiEUF9wbvO_p7fo9DE)

Thank you so much for being super awesome Josh!!! This was such an amazing experience. I look forward to more contributions

**Comment by kirkwaiblinger** [comment]
We'll also need a `no-duplicate-type-constitutents` version of this I think (see also https://github.com/typescript-eslint/typescript-eslint/issues/9344#issuecomment-2164770946)
